File: utils/filter_by_distance.py
import pandas as pd
import math
from process_csv import process_planitgeo_csv

def haversine(lat1, lon1, lat2, lon2):
    """
    Compute the great-circle distance between two points on Earth in miles.
    """
    R = 3958.8  # Earth radius in miles

    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlambda = math.radians(lon2 - lon1)

    a = math.sin(dphi/2)**2 + math.cos(phi1)*math.cos(phi2)*math.sin(dlambda/2)**2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

    return R * c


def properties_within_radius(df, center_lat, center_lon, radius_miles):
    if not {"Latitude", "Longitude"}.issubset(df.columns):
        raise ValueError("DataFrame must have 'Latitude' and 'Longitude' columns")

    def calc_distance(row):
        return haversine(center_lat, center_lon, row["Latitude"], row["Longitude"])

    # Compute distances and filter
    df["distance_miles"] = df.apply(calc_distance, axis=1)
    nearby_properties = df[df["distance_miles"] <= radius_miles]

    print(f"Found {len(nearby_properties)} properties within {radius_miles} miles.")
    return nearby_properties[["Primary ID", "Address", "Latitude", "Longitude", "distance_miles"]]


if __name__ == "__main__":
    csv_path = "../public/properties_data/planitgeo_properties_10_26.csv"
    csv_df = process_planitgeo_csv(csv_path)
    center_lat = 34.204689
    center_lon = -118.132674
    radius = 0.1  # miles

    nearby = properties_within_radius(csv_df, center_lat, center_lon, radius)
    print(nearby.head(34))

File: utils/filter_by_distance.ts
/**
 * Represents a property with location data
 */
export interface PropertyWithLocation {
  "Primary ID": string;
  Address: string;
  Latitude: number;
  Longitude: number;
  [key: string]: unknown; // Allow for additional fields
}

/**
 * Represents a property with calculated distance
 */
export interface PropertyWithDistance extends PropertyWithLocation {
  distance_miles: number;
}

/**
 * Compute the great-circle distance between two points on Earth in miles
 * using the Haversine formula.
 *
 * @param lat1 - Latitude of the first point in degrees
 * @param lon1 - Longitude of the first point in degrees
 * @param lat2 - Latitude of the second point in degrees
 * @param lon2 - Longitude of the second point in degrees
 * @returns Distance in miles
 */
export function haversine(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const R = 3958.8; // Earth radius in miles

  const phi1 = (lat1 * Math.PI) / 180;
  const phi2 = (lat2 * Math.PI) / 180;
  const dphi = ((lat2 - lat1) * Math.PI) / 180;
  const dlambda = ((lon2 - lon1) * Math.PI) / 180;

  const a =
    Math.sin(dphi / 2) ** 2 +
    Math.cos(phi1) * Math.cos(phi2) * Math.sin(dlambda / 2) ** 2;
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  return R * c;
}

/**
 * Filter properties within a specified radius from a center point.
 *
 * @param properties - Array of properties with location data
 * @param centerLat - Latitude of the center point
 * @param centerLon - Longitude of the center point
 * @param radiusMiles - Radius in miles
 * @returns Array of properties within the radius, with distance calculated
 * @throws Error if properties don't have required Latitude/Longitude fields
 */
export function propertiesWithinRadius(
  properties: PropertyWithLocation[],
  centerLat: number,
  centerLon: number,
  radiusMiles: number,
): PropertyWithDistance[] {
  // Validate that properties have required fields
  if (properties.length > 0) {
    const firstProperty = properties[0];
    if (!("Latitude" in firstProperty) || !("Longitude" in firstProperty)) {
      throw new Error("Properties must have 'Latitude' and 'Longitude' fields");
    }
  }

  // Calculate distances and filter
  const propertiesWithDistances: PropertyWithDistance[] = properties.map(
    property => ({
      ...property,
      distance_miles: haversine(
        centerLat,
        centerLon,
        property.Latitude,
        property.Longitude,
      ),
    }),
  );

  const nearbyProperties = propertiesWithDistances.filter(
    property => property.distance_miles <= radiusMiles,
  );

  console.log(
    `Found ${nearbyProperties.length} properties within ${radiusMiles} miles.`,
  );

  return nearbyProperties;
}

/**
 * Extract only relevant fields from properties with distance
 *
 * @param properties - Array of properties with distance
 * @returns Array of properties with only key fields
 */
export function extractKeyFields(properties: PropertyWithDistance[]): Array<{
  "Primary ID": string;
  Address: string;
  Latitude: number;
  Longitude: number;
  distance_miles: number;
}> {
  return properties.map(property => ({
    "Primary ID": property["Primary ID"],
    Address: property.Address,
    Latitude: property.Latitude,
    Longitude: property.Longitude,
    distance_miles: property.distance_miles,
  }));
}
